pyccl/halos/concentration.py

```diff
@@ -0,0 +1,308 @@
+        the corresponding mass definition that was used to initialize
+        this object.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            float or array_like: mass according to this object's
+            mass definition.
+        """
+        if mdef_other is not None:
+            M_use = mdef_other.translate_mass(cosmo, M, a, self.mdef)
+        else:
+            M_use = M
+        return M_use
+
+    def get_concentration(self, cosmo, M, a, mdef_other=None):
+        """ Returns the concentration for input parameters.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): the mass definition object
+                that defines M.
+
+        Returns:
+            float or array_like: concentration.
+        """
+        M_use = self._get_consistent_mass(cosmo,
+                                          np.atleast_1d(M),
+                                          a, mdef_other)
+
+        c = self.concentration(cosmo, M_use, a)
+        if np.ndim(M) == 0:
+            c = c[0]
+        return c
+
+
+class ConcentrationDiemer15(Concentration):
+    name = 'Diemer15'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationDiemer15, self).__init__(mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _setup(self):
+        self.kappa = 1.0
+        self.phi_0 = 6.58
+        self.phi_1 = 1.27
+        self.eta_0 = 7.28
+        self.eta_1 = 1.56
+        self.alpha = 1.08
+        self.beta = 1.77
+
+    def _check_mdef(self, mdef):
+        if (np.fabs(mdef.Delta - 200.) > 1E-4) and \
+           (mdef.rho_type != 'critical'):
+            return True
+        return False
+
+    def concentration(self, cosmo, M, a):
+        M_use = np.atleast_1d(M)
+
+        # Compute power spectrum slope
+        R = mass2radius_lagrangian(cosmo, M_use)
+        lk_R = np.log(2.0 * np.pi / R * self.kappa)
+        # Using central finite differences
+        lk_hi = lk_R + 0.005
+        lk_lo = lk_R - 0.005
+        dlpk = np.log(linear_matter_power(cosmo, np.exp(lk_hi), a) /
+                      linear_matter_power(cosmo, np.exp(lk_lo), a))
+        dlk = lk_hi - lk_lo
+        n = dlpk / dlk
+
+        sig = sigmaM(cosmo, M_use, a)
+        delta_c = 1.68647
+        nu = delta_c / sig
+
+        floor = self.phi_0 + n * self.phi_1
+        nu0 = self.eta_0 + n * self.eta_1
+        c = 0.5 * floor * ((nu0 / nu)**self.alpha +
+                           (nu / nu0)**self.beta)
+        if np.ndim(M) == 0:
+            c = c[0]
+
+        return c
+
+
+class ConcentrationBhattacharya13(Concentration):
```

Comment:
Each subclass will need a class-level doc string. Otherwise readthedocs will be ugly. Copy-paste is the best!

Reply:
Done!